### ISSUE TITLE
Provisioner sudo sub-shell commands are not fully evident in the final template

### ISSUE THREAD

**Issue by rmwiseman** [hashicorp/packer#5396]
Apologies for the rubbish title, but it's really hard to describe what has been happening!  I've experienced a very weird issue that appears to be to do with my use of a sub-shell within a provisioner.  The short version is that what I see happen in a `sudo -i -u `_`username`_ sub-shell all works but the effects are not evident in the template.

I'm installing Terraform (using a Packer provisioner) as a different user (called `terraform`) using the `sudo` construct:

    sudo -i -u terraform <<'EOF'
    # stuff to perform as the terraform user
    EOF

The issue I've had is that if this is the last thing in the last provisioner then what was done in the sub-shell is not properly saved.  Within the povisioner, I run `ls -l --full-time /home/terraform/bin/terraform` and the output is:

    -rwxrwxr-x. 1 terraform terraform 55308544 2017-09-19 14:57:50.000000000 +0100 /home/terraform/bin/terraform

When the template has been created and I spin up an instance of it, I run exactly the same command as the same user and get:

    -rwxrwxr-x. 1 terraform terraform 0 2017-09-19 14:57:50.000000000 +0100 /home/terraform/bin/terraform

So it's identical except that the file is now empty!  There's nothing in my provisioner that could empty the file.

Similarly, I modify `~terraform/.bashrc` by appending a line.  Running `ls -l --full-time /home/terraform/.bashrc` within the provisioner (after the line has been appended) gives:

    -rw-r--r--. 1 terraform terraform 259 2017-09-28 08:24:50.294575178 +0100 /home/terraform/.bashrc

and the same command run by the same user in an instance of the template gives:

     -rw-r--r--. 1 terraform terraform 231 2017-09-28 08:24:50.294575178 +0100 /home/terraform/.bashrc

This is perhaps odder because the timestamp is exactly the same (indicating that the file was modified) but the appended line, which is:

    export PATH=".:~/bin:$PATH"

(28 bytes long including the newline) is missing from the template.

**Simplest case:**

As requested, I've produced a simplest case example.  The provisioner is:

```
  "provisioners": [
    {
      "type": "shell",
      "inline": [
        "set -x",
        "sudo adduser fred",
        "echo 'C-pZ&z17$m(=\nC-pZ&z17$m(=' | sudo passwd fred 2>/dev/null",
        "sudo usermod -aG wheel fred",
        "sudo -i -u fred <<'EOF'",
        "set -x",
        "ls -l --full-time newfile.txt",
        "echo Just testing > newfile.txt",
        "ls -l --full-time newfile.txt",
        "EOF"
      ]
    }
  ]
```

The first `ls` is just intended to indicate that `newfile.txt` does not yet exist.

The output of the second `ls` within the provisioner is:

    -rw-rw-r--. 1 fred fred 13 2017-09-28 09:10:41.282463064 +0100 newfile.txt

but when run on an instance created from the template, the output is:

    -rw-rw-r--. 1 fred fred 0 2017-09-28 09:10:41.282463064 +0100 newfile.txt

The terminal output is here: https://gist.github.com/rmwiseman/b539365e543459c0d5e1c35157e42750

The log file (`packer-log019546346`), which is basically the same, is here: https://gist.github.com/rmwiseman/c112b6969a3c21a4087ab270b627ac5c

And the `bug.json` script (credentials removed) is here: https://gist.github.com/rmwiseman/60d0749bee00ffaa8c4bdb3757ab7383

**Workaround:**

If I add a `sleep 10` _after_ the provisioner's sub-shell completes, then it all works as it should and the instance spun up from the template contains files with sizes consistent with those shown during provision.

**Extra information:**

I'm using Packer 1.1.0 provisioning on a CentOS 7 VM (specifically 7.4.1708 after a `yum update`) for provision, running on CloudStack 4.5.2.2.

**Comment by rickard-von-essen** [comment]
Try doing a `sync` instead of the `sleep`. It seems to be an issue with that the snapshot is done before your data has landed on the disk.

**Comment by rmwiseman** [comment]
Thanks @rickard-von-essen, I'm trying that now.  It's bizarre to think that you can successfully run a command before the command has been written to persistent storage though!  I'll update you when it has completed...

**Comment by mwhooker** [comment]
we generally try not to be too clever with packer. I could imagine mentioning this in the docs, though

**Comment by rmwiseman** [comment]
@mwhooker I definitely think it should at least be prominent in a FAQ or troubleshooting guide!  It had me confused for ages.

One thing I did wonder was instead of "being clever" by running `sync` on the VM before templating it, perhaps all Packer needs to do to mitigate this problem is pause for a couple of seconds after the provisioning is complete.  I imagine that would have the same effect as my `sleep` at the end of the provisioner.  I realise that would make all provisioning actions take two seconds longer, but perhaps that's a small price to pay to avoid confusion.  It could even be configurable: you can remove the pause to speed up your script, but the documentation for the option would then explain that you might have to do a `sync` at the end of your provisioning.

**Comment by rickard-von-essen** [comment]
> perhaps all Packer needs to do to mitigate this problem is pause for a couple of seconds after the provisioning is complete

Unfortunately there is no guarantee that this will help.

**Comment by rmwiseman** [comment]
Surely there must be something Packer can do to ensure cached data is written to disk before the template is created??

**Comment by rickard-von-essen** [comment]
> Surely there must be something Packer can do to ensure cached data is written to disk before the template is created

Not from the OS side. But there is something fishy here. The builder shuts down the instance that should properly sync all fs when the OS shuts down.

@svanharmelen Do you know if https://github.com/hashicorp/packer/blob/master/builder/cloudstack/step_shutdown_instance.go#L31 waits for the instance to shutdown completely or are we missing a wait for state here?

**Comment by svanharmelen** [comment]
Yes... We use the [async client](https://github.com/hashicorp/packer/blob/master/builder/cloudstack/builder.go#L38) which waits until the call is finished by polling the running async job. So the builder blocks here until CloudStack confirms that the stop call is finished and the VM is stopped.

**Comment by rmwiseman** [comment]
As you say, @rickard-von-essen there's definitely something fishy going on!

Is there a perhaps a call that can/should be done before the stop call?

Or could this actually be a CloudStack issue?

**Comment by svanharmelen** [comment]
> Is there a perhaps a call that can/should be done before the stop call?

The `stop` call, is the call this signals the OS to shutdown and waits until it's stopped before stopping the VM itself on the hypervisor. So the answer to that question would be no.

> Or could this actually be a CloudStack issue?

Maybe... Or maybe it's and hypervisor thingy... Or maybe something in your OS... Hard to say. I'll see if I can find some time to test your issue on our installation and see if I can reproduce it. We're using the latest CS version on KVM hypervisors. Will let you know the results when I got them.

**Comment by rmwiseman** [comment]
Thanks @svanharmelen.  I stumbled across a mention of this [in the Packer documentation](https://www.packer.io/docs/builders/virtualbox-iso.html#basic-example), actually, where `shutdown_command` is mentioned.  This was for VirtualBox, though, not CloudStack.  I [googled](http://www.google.com/search?q=site:packer.io+shutdown_command+sync) to see if it's mentioned for other provisionsers and it is -- just not for CloudStack.  Should `shutdown_command` be in the CloudStack provisioner too, or is this a related but different issue?

**Comment by svanharmelen** [comment]
@rmwiseman this is not related.

All the builders that the google search returned are "stand alone" virtualisation solutions that are not being managed or orchestrated.

Clouds like AWS, Azure, Google, OpenStack and also CloudStack orchestrate and manage your workload (VM's) on their hypervisors. So if you shutdown a VM from within the OS, these orchestrators will start the VM right back up again as they didn't know that you manually triggered the shutdown from within the OS and thus believe it still needs to be in a running state.

So in these clouds, you need to tell the orchestrator you want to stop a VM in order to stop it. The Orchestrator will then signal the OS to shutdown and once that is done the VM itself is deallocated from the hypervisor.

**Comment by rmwiseman** [comment]
Ah OK, thanks for the clarification @svanharmelen.  I just thought I should check!

**Comment by svanharmelen** [comment]
@rmwiseman I tried to reproduce it this morning, but I was unable to do so...

```
> packer build bla.json
cloudstack output will be in this color.

==> cloudstack: Preparing config...
    cloudstack: Config has been prepared!
==> cloudstack: Creating instance...
    cloudstack: Instance has been created!
==> cloudstack: Setup networking...
    cloudstack: Associating public IP address...
    cloudstack: Creating port forward...
    cloudstack: Creating network ACL rule...
    cloudstack: Networking has been setup!
==> cloudstack: Waiting for SSH to become available...
==> cloudstack: Connected to SSH!
==> cloudstack: Provisioning with shell script: /var/folders/jm/knykn8sd6vx98n14gfj2h08r6k_qt_/T/packer-shell587143338
    cloudstack: + sudo adduser fred
    cloudstack: + echo 'C-pZ&z17$m(=
    cloudstack: C-pZ&z17$m(='
    cloudstack: + sudo passwd fred
    cloudstack: + sudo usermod -aG wheel fred
    cloudstack: Changing password for user fred.
    cloudstack: + sudo -i -u fred
    cloudstack: passwd: all authentication tokens updated successfully.
    cloudstack: + ls -l --full-time newfile.txt
    cloudstack: ls: cannot access newfile.txt: No such file or directory
    cloudstack: + echo Just testing
    cloudstack: + ls -l --full-time newfile.txt
    cloudstack: -rw-rw-r--. 1 fred fred 13 2017-10-03 10:53:34.472306000 +0200 newfile.txt
==> cloudstack: Shutting down instance...
    cloudstack: Instance has been shutdown!
==> cloudstack: Creating template: Bug on CentOS
    cloudstack: Retrieving the ROOT volume ID...
    cloudstack: Creating the new template...
    cloudstack: Template has been created!
==> cloudstack: Cleanup networking...
    cloudstack: Deleting network ACL rule...
    cloudstack: Deleting port forward...
    cloudstack: Releasing public IP address...
    cloudstack: Networking has been cleaned!
==> cloudstack: Deleting instance...
    cloudstack: Instance has been deleted!
Build 'cloudstack' finished.

==> Builds finished. The artifacts of successful builds are:
--> cloudstack: A template was created: Bug on CentOS
```

The output of the build itself looks identical to yours. But when I create a VM of the newly created template and login, in my case it shows the file and it's content correctly:

```
[bootstrap@svhtest01 ~]$ sudo su - fred
[fred@svhtest01 ~]$ ls -l
total 4
-rw-rw-r--. 1 fred fred 13 Oct  3 10:53 newfile.txt
[fred@svhtest01 ~]$ cat newfile.txt
Just testing
```

Of course my source template is a different one, but it's also a Centos 7 template:

```
[fred@svhtest01 ~]$ cat /etc/centos-release
CentOS Linux release 7.4.1708 (Core)
```

So it seems the logic used in the CloudStack builder is OK... I asked our internal cloud team (who are big contributors of CloudStack and Cosmic) and they didn't recall anything like this. It seems unlikely it's CloudStack related. They would think more into the direction of the hypervisors...

**Comment by rmwiseman** [comment]
@svanharmelen Thanks for trying it out.  Strangely enough, a colleague of mine had the same problem (although not, as far as I know, related to a `sudo` sub-shell) this morning: users were created but their `~/.bashrc` and `~/.bash_profile` files were (in the instance) empty.  Possibly unrelatedly, some `.lock` files still existed that shouldn't have, too.

Is there anything else I can provide that might help get to the bottom of this?

**Comment by SwampDragons** [comment]
Since I don't think there's anything Packer can directly do about this, I am going to close for now. If we end up getting more information that's actionable from Packer's side, we can reopen or open a new issue. I would like to hear if you end up finding any resolution, though.
